CA.pl thorough structure change

Tatsuya-Nonogaki opened the pull request:
Thorough structure change for ease of comprehension and function reuse. Plus some bug fixes.

richsalz commented:
this looks nice, but it's really a big change for late in the release.  perhaps if @levitte agrees...

levitte commented on apps/CA.pl.in:
```diff
@@ -77,6 +91,19 @@ sub copy_pemfile
     return $found == 2 ? 0 : 1;
 }
 
+sub copy_any
```
Quite frankly, I'd rather see File::Copy being used.  It's been with perl all the way since 5.002, so it's safe to use.

Tatsuya-Nonogaki replied:
It obviously shortens the codes. But I wonder if it is reasonable to load File::Copy module for a few utilization...

levitte replied:
I find it a little more reasonable than re-inventing the wheel yet again...

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,231 @@ if ($WHAT eq '-newcert' ) {
+}
+
+# Update Certificate Revocation List
+sub crl
```
Please call it `gencrl`

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,231 @@ if ($WHAT eq '-newcert' ) {
+}
+if ($WHAT eq '-newcert' ) {
+} elsif ($WHAT eq '-newreq' ) {
+    &newreq();
```
We'd rather see function calls without `&` unless it's absolutely necessary

levitte commented on apps/CA.pl.in:
```diff
@@ -7,39 +7,53 @@
-my $PKCS12 = "$openssl pkcs12";
+our $OPENSSL_CONFIG = defined $ENV{'OPENSSL_CONFIG'} ? $ENV{'OPENSSL_CONFIG'} : "";
+our $OPENSSL_CONFIG_PAIRBAG = "-config /etc/pki/tls/openssl-pairbag.cnf";
+our $CERTTOP = defined $ENV{'CERTTOP'} ? $ENV{'CERTTOP'} : "/etc/pki/pairbag";
```
I have no idea what the pairbag, but it seems to be something out of scope for OpenSSL...

levitte commented:
This might be a worthy refactoring, but it's way too late for added features, so the pairbag stuff has to go.

I'll do another read through tomorrow, when a bit more awake than I am right now...

mattcaswell commented:
Ping @levitte. Are we progressing this for 1.1.0 or not?

levitte commented:
This hasn't moved at all since my remarks, I think will have to be a post 1.1.0 refactor.

Tatsuya-Nonogaki commented:
Thanks for all your discussions. I made some brush-ups regarding their comments - gencrl() and "&" stuff, in my own clone repository. So what can I do?

mattcaswell commented:
@Tatsuya-Nonogaki, just push the new commits to this branch on github and the PR will be updated with them. Put a note in the comments so that @levitte can take a look.

Tatsuya-Nonogaki commented:
I wrongly submit a pull request to openssl/master, not here. I was mistaken. Please ignore it.

richsalz commented:
ping @levitte what do you think?  See #1797

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+        $days = $days. " ". $day;
+    } elsif ($days !~ /^-days[[:blank:]]+\w+/) {
+        $days = "";
+    }
```
It's quite unclear that you're actually getting `@ARGV` passed through `@_` and that this code is actually parsing command line arguments.  You *could* do the latter in a more readable form:

``` perl
    my $days = "";

    while (scalar @_ > 0) {
        if ($_[0] eq '-days') {
            die "-days must have a value\n" unless scalar @_ > 1;
            $days = join(" ", shift @_, shift @_);
        } else {
            last;
        }
    }
    die "extra arguments to xsign\n" unless scalar @_ == 0;
```

Tatsuya-Nonogaki replied:
It is not generally good to believe user input though this is not a web cgi. I don't want to let arbitrary number of arguments injected into sub routines. Anyway, excessive arguments must be discarded somewhere through. I do it right at the entrance.
In addition to it, when #1797 `-extra` extension is ported also into this milestone, circumstance may change. I think this style is considerably the best for now. I am going to examine how they looks like with #1797 later.

**About '-days xxx' argument handling:**
No, it is not so simple. `-days` arguments come into these sub routines via two way.
1. As arguments to CA.pl
2. Called internally from other subs or main scope
In case 1, a user will execute `CA.pl -xsign -days 730`, my `$days` gets '`-days`' and `$day` '`730`'. There is a chance he might unusually enclose the whole arguments in a pair of qtuotes, in that case `$days` gets "`-days 730`" and `$day` undef.
In case 2, we almost always pass `-days` arguments in a form of a variable `$DAYS`. Result is the same as the latter in case 1. So they converge onto three patterns:
1. $days="-days" $day="_xxx_"
2. $days="-days" $day=undef (erroneously)
3. $days="-days (extra spaces could be) _xxx_" $day=undef
I tested them all and came to this conclusion. I took just the `join()` style.
Just for record, these subs with `-days` expect two arguments **at most**. I don't see a reason to pass whole `@ARGV` to them.

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+        $days = $days. " ". $day;
+    } elsif ($days !~ /^-days[[:blank:]]+\w+/) {
+        $days = "";
+    }
```
Same comment on argument parsing as for `xsign`

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+        $days = $days. " ". $day;
+    } elsif ($days !~ /^-days[[:blank:]]+\w+/) {
+        $days = "";
+    }
```
Same command line parsing comment here

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+        $days = $days. " ". $day;
+    } elsif ($days !~ /^-days[[:blank:]]+\w+/) {
+        $days = "";
+    }
```
Same command line parsing argument...

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+# Verify certificates against CA certificate for CATOP 
+sub verify
+{
     my @files = @ARGV ? @ARGV : ( $NEWCERT );
```
Why don't you pass `@ARGV` through `@_` like you did in most other functions?

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+} elsif ($WHAT eq '-selfsign' ) {
+    selfsign($ARGV[0],$ARGV[1]);
+} elsif ($WHAT eq '-verify' ) {
+    verify();
```
You should pass `@ARGV` here

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+} elsif ($WHAT eq '-revoke' ) {
+    revoke($ARGV[0],$ARGV[1]);
+} elsif ($WHAT eq '-p12mng-revoke' ) {
+    p12mng_revoke($ARGV[0],$ARGV[1]);
```
In all these cases, I think it would be better to pass `@ARGV` instead of a selection of it, and properly parse the arguments in each function.

levitte commented:
Regarding the p12mng stuff, it would really need an explanation.  Is it connected to some other software (the `/etc/pki/` path seems to indicate something like that)?  Why does it need an entirely different directory structure?  Why make it `/etc/pki/...` by default instead of, say, `./p12mng/...`?

Either way, `docs/mn1/CA.pl.pod` needs to be updated with the p12mng stuff explained, and `test/recipes/80-test_ca.t` probably needs an update as well to cover the p12mng stuff (if you need help with the test stuff, please tell me.  I'll be happy to lend a hand)

Tatsuya-Nonogaki commented:
I will test and consider them. I will be busy this weekend. Please wait a while.

Tatsuya-Nonogaki commented on apps/CA.pl.in:
```diff
@@ -6,40 +6,58 @@
+use File::Copy;
 
-my $openssl = "openssl";
+our ($ME) = ($0 =~ m|/([^/]+)$|);
```
It has been applied.

Tatsuya-Nonogaki commented on apps/CA.pl.in:
```diff
@@ -6,40 +6,58 @@
-my $PKCS12 = "$openssl pkcs12";
+our $OPENSSL_CONFIG = defined $ENV{'OPENSSL_CONFIG'} ? $ENV{'OPENSSL_CONFIG'} : "";
+our $OPENSSL_CONFIG_P12MNG = "-config /etc/pki/tls/openssl-p12mng.cnf";
+our $CERTTOP = defined $ENV{'CERTTOP'} ? $ENV{'CERTTOP'} : "/etc/pki/p12mng";
```
It has been applied.

Tatsuya-Nonogaki commented on apps/CA.pl.in:
```diff
@@ -6,40 +6,58 @@
+our $X509 = "$openssl x509";
+our $PKCS12 = "$openssl pkcs12";
+our $CNAMETXT = "My Certificate";
+our $EXPORTPASS_FILE = "exportpw";
```
They have been changed.

Tatsuya-Nonogaki commented on apps/CA.pl.in:
```diff
@@ -6,40 +6,58 @@
-my $RET = 0;
-my $WHAT = shift @ARGV;
-my $FILE;
+our $CATOP = defined $ENV{'CATOP'} ? $ENV{'CATOP'} : "./demoCA";
```
Applied.

Tatsuya-Nonogaki commented on apps/CA.pl.in:
```diff
@@ -59,21 +77,30 @@ sub crl_reason_ok
+    $outmode = ">" unless $outmode;
+    if ($infile eq $outfile) {
+        $onthefly = 1;
+        copy($infile,$infile."_tmp") || die "Cannot copy $infile, $!";
```
Sorry, I don't really get your point.

levitte replied:
That was, considering `$outmode` is always `">"`.  I totally spaced on the fact that `$outmode` could be `">>"`...  so, never mind

Tatsuya-Nonogaki commented on apps/CA.pl.in:
```diff
@@ -87,32 +114,40 @@ sub run
-if ($WHAT eq '-newcert' ) {
-    # create a certificate
-    $RET = run("$REQ -new -x509 -keyout $NEWKEY -out $NEWCERT $DAYS");
+# Create a certificate
```
Corrected.

Tatsuya-Nonogaki commented on apps/CA.pl.in:
```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+            . " $days -infiles $NEWREQ");
+}
+
+# Same as above except the certificate is writen to STDOUT
```
Corrected.

Tatsuya-Nonogaki commented:
To "Regarding the p12mng stuff.." question from levitte:
(Sorry for a long comment)
I created `p12mng_new` sub routine for one of my business client to manage browser certificates for public-key authentication to their Apache web site which I deployed. This helps the not-much-skilled personel adding and revoking certificates daily. With the dedicated directory structure, certificates are created there whatever current directory he is in. And it is easy to retrieve, maintain, backup hundreds of certs. It can't be precisely controled only by openssl.cnf nor CA.pl, because most definitions(variables) are independent from each other. If we heavily rely on one of the two, each purpose needs indivisual CA directory. With current framework, shared or separate CA is your choice.

Later on, I made a series of subs to prepare vSphere vCenter and ESXi 5.x/6.0 keys/certificates signed with private CA. They are too specific to contribute here. vSphere certs management is really a beast. For example, a single vCenter needs four to seven set of cert/key/chain, per service. We need to generate new ESXi certificates when a host machine is added. In this case, too, dedicated directory is worthy. The sub routines were rather easily made by copy/paste p12mng stuff. I proposed p12mng here not only since it can be used directly, but also with a hope that it will be a base of custom routines like I did.

One thing I am worried about is p12mng may have to be accompanied by an example openssl-p12mng.cnf. I can serve one at once though it must be examined by OpenSSL professionals. What part of the tree should it be slipped in?

levitte commented on apps/CA.pl.in:
```diff
@@ -127,59 +163,289 @@ if ($WHAT eq '-newcert' ) {
+    # remove the revoked key set
+    run("rm -i ${CERTTOP}/certs/${cname}.pem");
+    run("rm -i ${CERTTOP}/private/${cname}.key");
+    run("rm -i ${CERTTOP}/private/${cname}.p12");
```
These removals aren't a good idea, generally speaking.  Revoked certificates can still be used to validate signatures, if that's what they were used for.

levitte commented:
Ok, I've read through all this and your explanations quite a few times, and I do understand where you're coming from...  however, it seems to me like your changes modify CA.pl from a general purpose demo tool to, well, something else, and specific to *your* purpose more than something of general interest (I agree that's arguable).

Having looked quite a few times at the suggested changes to `$CATOP` and a few more variables, I realise that these changes as a whole can't be made into OpenSSL before version 1.2.0 as they currently stand.  They change CA.pl's behavior in an incompatible way.

As for the p12mng stuff and the changes to use specific /etc paths, you could just as easily do that in a separate script that merely uses CA.pl as it currently stands.  I'd say this is preferable unless it can be shown that this *must* be part of CA.pl.

That being said, I *like* the structural change of CA.pl into subs for each subcommand, so essentially, if the p12mng stuff and directory structure changes can be separated out from this PR, I will be much happier.

Tatsuya-Nonogaki commented:
I honestly admit `p12mng_*` have been sitting uncomfortably in the file, out of place in short. Now CA.pl's back to basics with some improvements made along the way. I cleaned it up, added three `show*` subs. I have been feeling they should be there if it is 

> intended as a simple front end for the openssl program for use by a beginner

but also it saves a bit of my keystrokes.

openssl-machine commented:
This PR has been closed.  It was waiting for the creator to make requested changes but it has not been updated for 305 days.
